src/Client/Updater.php

```diff
@@ -105,13 +98,77 @@ public function validateTarget($target_repo_path, $target_stream)
         $timestamp_structure = json_decode($timestamp_contents, true);
         // SPEC: 2.1
         if (! $this->checkSignatures($timestamp_structure, 'timestamp')) {
-          // Exception? Log + return false?
             throw new \Exception("Improperly signed repository timestamp.");
         }
 
+        // SPEC: 2.2
+        $currentStateTimestamp = json_decode($durableStorage['timestamp.json'], true);
+        $this->checkRollbackAttack($currentStateTimestamp['signed'], $timestamp_structure['signed']);
+
+        // SPEC: 2.3
+        $this->checkFreezeAttack($timestamp_structure['signed'], $now_date);
+        $durableStorage['timestamp.json'] = $timestamp_contents;
+
         return true;
     }
 
+    protected function metadataTimestampToDatetime(string $timestamp) : \DateTimeImmutable
+    {
+        return \DateTimeImmutable::createFromFormat("Y-m-d\TH:i:sT", $timestamp);
+    }
+
+    /**
+     * Verifies that an incoming remote version of a metadata file is >= the last known version.
+     *
+     * @param $localMetadata
+     * @param $remoteMetadata
+     * @throws RollbackAttackException
+     */
+    protected function checkRollbackAttack(array $localMetadata, array $remoteMetadata)
+    {
+        $localVersion = $localMetadata['version'] + 0;
+        if ($localVersion == 0) {
+            // Failsafe: if local metadata just doesn't have a version property or it is not an integer,
+            // we can't perform this check properly.
+            throw new RollbackAttackException("Empty or invalid local timestamp version \"${localMetadata['version']}\"");
+        }
+        if ($remoteMetadata['version'] + 0 < $localVersion) {
+            throw new RollbackAttackException("remote timestamp metadata version \"${remoteMetadata['version']}\" is less than previously seen timestamp version \"${localMetadata['version']}\"");
+        }
+    }
+
+    /**
+     * Verifies that metadata has not expired, and assumes a potential freeze attack if it has.
+     *
+     * @param array $metadata
+     * @param \DateTimeInterface $now
+     * @throws FreezeAttackException
+     */
+    protected function checkFreezeAttack(array $metadata, \DateTimeInterface $now)
+    {
+        $metadataExpiration = $this->metadataTimestampToDatetime($metadata['expires']);
+        $type = '';
+        if (!empty($metadata['_type'])) {
+            $type = $metadata['_type'];
+        }
+        if ($metadataExpiration < $now) {
+            throw new FreezeAttackException(sprintf("remote %s metadata expired on %s", $type, $metadataExpiration->format('c')));
+        }
+    }
+
+  /**
+   * Validates a target.
+   *
+   * @param $target_repo_path
+   * @param $target_stream
+   *
+   * @return bool
+   *   Returns true if the target validates.
+   */
+    public function validateTarget($target_repo_path, $target_stream)
+    {
+    }
```

Comment:
I thought we chatted about this and agreed the todo wasn't necessary here since we wouldn't forget but I see this is still open so I created an issue. Perhaps I misunderstood the conclusion of that conversation. Resolving.